eth/filters/api.go

```diff
@@ -203,15 +201,14 @@ func (s *PublicFilterAPI) NewPendingTransactionFilter() (string, error) {
 
 // newLogFilter creates a new log filter.
 func (s *PublicFilterAPI) newLogFilter(earliest, latest int64, addresses []common.Address, topics [][]common.Hash, callback func(log *vm.Log, removed bool)) (int, error) {
-	s.logMu.Lock()
-	defer s.logMu.Unlock()
-
 	filter := New(s.chainDb)
-	id, err := s.filterManager.Add(filter, LogFilter)
+	id, err := s.filterManager.Add(filter, LogFilter) // use inside logMu deadlocks filterLoop()
```

Comment:
Thanks a lot, you mean this could result into a read/write race here:
https://github.com/ethereum/go-ethereum/blob/develop/eth/filters/filter_system.go#L166 ?

That should be changed the same way that Remove works then, I guess. The comment holds, if Add() is called while logMu is held, then because Add() tries for sf.filterMu, it can deadlock filterLoop which can through the callbacks try the same locks in reverse order.